### DISCUSSION TITLE
Redesign MockConvas

### DISCUSSION THREAD

**Discussion by st-pasha** [flame-engine/flame#920] [Ideas]
Currently `MockCanvas` serializes all method calls into strings and stores them all in an internal list. That list is later tested to see if it contains "correct" strings. Something like this:
```dart
      final calls = canvas.methodCalls;
      expect(calls.length, 6);
      expect(
        calls[0],
        'transform(1.0, 0.0, 0.0, 0.0, -0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 1.0, 0.0, 23.0, 17.0, 0.0, 1.0)',
      );
      expect(calls[1], 'drawRect(0.0, 0.0, 10.0, 10.0)');
      expect(calls[2], 'drawLine(0.0, -2.0, 0.0, 2.0)');
      expect(calls[3], 'drawLine(-2.0, 0.0, 2.0, 0.0)');
      expect(calls[4], 'drawParagraph(?, -30.0, -15.0)');
      expect(calls[5], 'drawParagraph(?, -20.0, 10.0)');
```
This has several disadvantages, the chief among which is numeric instability. For example, the string `"0.0"` is different from `"-0.0"`, though both represent exactly the same numeric quantity. In general, the only way to write a more-or-less reliable test then is to make sure to always use very round numbers (like in the example above), and even that does not always work perfectly.

### Suggestion
I suggest then changing the internal structure of `MockCanvas` so that instead of storing strings it would store individual drawing commands, and provide the `==` operator which would compare canvases up to certain numeric precision. So, the usage code could become something like this:
```dart
final canvas = MockCanvas();
// draw something on canvas here
// ...
// now test that the correct drawing commands were issued:
expect(
  canvas,
  MockCanvas()
    ..translate(23, 17)
    ..drawRect(0, 0, 10, 10)
    ..drawLine(0, -2, 0, 2)
    ..drawLine(-2, 0, 2, 0)
    ..drawParagraph(-30, -15)
    ..drawParagraph(-20, 10),
);
```

**Comment by spydon** [comment]
Sounds great, it will be much easier to read what is happening too.

**Comment by luanpotter** [comment]
I love it! One caveat is to have a nice to way to do some sort of "contains actions" if you don't wanna check the whole sequence of actions on some tests.

**Comment by erickzanardo** [comment]
Sounds very good!

**Comment by spydon** [comment]
Do you want to take on this task @st-pasha? :)

**Reply by st-pasha** [reply]
sure, just convert this discussion into an issue and assign
